benchmarks/benchmark_serving.py

```diff
@@ -0,0 +1,589 @@
+    """
+    sampled_requests = []
+    for _ in range(num_requests):
+        prompt_len = int(np.random.normal(input_len_mean, input_len_stddev))
+        output_len = int(np.random.normal(output_len_mean, output_len_stddev))
+
+        # generate random id list for the prompt having length prompt_len
+        def gen_prompt_ids(prompt_len):
+            ids = []
+            for _ in range(prompt_len):
+                ids.append(random.choice(list(tokenizer.get_vocab().values())).value)
+            return ids
+
+        # Generte random prompt from tokenizer's vocabulary
+        prompt = tokenizer.decode(gen_prompt_ids(prompt_len), return_tensors="pt")
+        sampled_requests.append((prompt, prompt_len, output_len))
+    return sampled_requests
+
+
+async def get_request(
+    input_requests: List[Tuple[str, int, int]],
+    request_rate: float,
+) -> AsyncGenerator[Tuple[str, int, int], None]:
+    """
+    Asynchronously generates requests based on the input_requests and request_rate.
+
+    Args:
+        input_requests (List[Tuple[str, int, int]]): A list of input requests, where each request is a tuple
+            containing a string, an integer, and another integer.
+        request_rate (float): The rate at which requests should be generated. If set to float("inf"),
+            requests will be generated without any delay.
+
+    Yields:
+        Tuple[str, int, int]: A request tuple containing a string, an integer, and another integer.
+
+    """
+    for request in input_requests:
+        yield request
+
+        if request_rate == float("inf"):
+            # If the request rate is infinity, then we don't need to wait.
+            continue
+        # Sample the request interval from the exponential distribution.
+        interval = np.random.exponential(1.0 / request_rate)
+        # The next request will be sent after the interval.
+        await asyncio.sleep(interval)
+
+
+async def send_request(
+    api_url: str,
+    prompt: str,
+    prompt_len: int,
+    output_len: int,
+    config: dict,
+    track_token_latency: bool = True,
+    track_input_output: bool = False,
+    progress_bar: tqdm = None,
+) -> None:
+    """
+    Sends a request to the specified API URL with the given prompt and configuration.
+
+    Args:
+        api_url (str): The URL of the API to send the request to.
+        prompt (str): The prompt text.
+        prompt_len (int): The length of the prompt text.
+        output_len (int): The desired length of the output.
+        config (dict): The configuration for the request.
+        progress_bar (tqdm, optional): A progress bar to update during the request. Defaults to None.
+    """
+    request_start_time = time.perf_counter()
+
+    headers = {"User-Agent": "Benchmark Client"}
+
+    # Use sample output_len if max_new_tokens not specified
+    if "max_new_tokens" in config:
+        output_len = config["max_new_tokens"]
+    else:
+        config["max_new_tokens"] = output_len
```

Comment:
Or make a deep copy of the config parameters here. Otherwise, when passing through line 231 for the first time, the value of `config` will become {'max_new_tokens': 37}, so that the output_len will be 37 every time in the future. This is why total_time is reduced a lot

Reply:
good catch, pls feel free to fix this.